scripts/circleci/upload_circleci_artifacts.py

```diff
@@ -0,0 +1,404 @@
+    from typing import Generator
+    from typing import Tuple
+
+import argparse
+import os
+import itertools
+import operator
+import datetime
+import time
+import re
+import json
+
+import requests
+
+CIRCLE_API_URL = "https://circleci.com/api/v2"
+
+CIRCLE_API_PROJECT_URL = CIRCLE_API_URL + "/project/gh/scalyr/scalyr-agent-2"
+
+# 15 min
+CIRCLE_WAIT_TIMEOUT = 60 * 15
+
+try:
+    CIRCLE_API_TOKEN = compat.os_environ_unicode["CIRCLE_API_TOKEN"]
+except KeyError:
+    print("Environment variable 'CIRCLE_API_TOKEN' is not specified.")
+    raise
+
+
+def _do_request(method, url, **kwargs):
+    # type: (str, str, **Any) -> requests.Response
+
+    headers = kwargs.get("headers", dict())
+
+    headers["Circle-Token"] = CIRCLE_API_TOKEN
+    headers["Accept"] = "application/json"
+
+    kwargs["headers"] = headers
+
+    with requests.Session() as session:
+        resp = session.request(method=method, url=url, **kwargs)
+    resp.raise_for_status()
+    return resp
+
+
+def get_request(url, **kwargs):
+    # type: (str, Dict) -> Dict
+    resp = _do_request("GET", url, **kwargs)
+    return resp.json()
+
+
+def post_request(url, **kwargs):
+    # type: (str, Dict) -> requests.Response
+    return _do_request("POST", url, **kwargs)
+
+
+def get_paginated_collection(url):
+    # type: (str) -> Generator
+    """
+    Generator witch fetches elements from circleci collection with pagination.
+    :param url: Url to the collection.
+    :return: Yields next element in the collection.
+    """
+    next_page_token = None
+    while True:
+
+        resp = get_request(url=url, params={"page-token": next_page_token})
+
+        for item in resp["items"]:
+            yield item
+
+        next_page_token = resp["next_page_token"]
+        if next_page_token is None:
+            raise StopIteration()
+
+
+def get_paginated_list(url):
+    # type: (str) -> List
+    """
+    Fetch the circleci paginated collection as list.
+    """
+
+    return list(get_paginated_collection(url))
+
+
+def get_paginated_dict(url, key):
+    # type: (str, str) -> Dict
+    """
+    Fetch the circleci paginated collection as dict.
+    :param key: Name of the field of the element which will be used as key for the dictionary.
+    """
+    return {item[key]: item for item in get_paginated_collection(url)}
+
+
+def download_artifact_file(artifact_info, output_path):
+    # type: (Dict, str) -> None
+    """
+    Download circleci job artifact.
+    :param artifact_info: Contains information about artifact.
+    :param output_path: Base output path.
+    :return:
+    """
+    artifact_output_path = os.path.join(
+        output_path, os.path.basename(artifact_info["path"]),
+    )
+
+    with requests.Session() as session:
+        resp = session.get(url=artifact_info["url"], allow_redirects=True, stream=True)
+
+    with open(artifact_output_path, "wb") as file:
+        for chunk in resp.iter_content(chunk_size=8192):
+            file.write(chunk)
```

Comment:
I wanted to say that it would be good to verify the checksum of the downloaded file to make sure we are working with a correct and non-corrupted file, but it looks like Circle CI API (https://circleci.com/docs/api/#artifacts) doesn't expose the checksum.

---

EDIT: I had a look and it looks the the server also doesn't return E-Tag or a similar header so there isn't much we can do (option would be to to publish ``<file name>.sha256`` file as part of the build and use that, but this may be an overkill and not needed for now).